nucliadb/nucliadb/purge/orphan_shards.py

```diff
@@ -0,0 +1,198 @@
+    stored_shards: dict[str, ShardLocation] = {}
+    shards_manager = KBShardManager()
+
+    async with driver.transaction(read_only=True) as txn:
+        async for kbid, _ in KnowledgeBox.get_kbs(txn, slug=""):
+            try:
+                kb_shards = await shards_manager.get_shards_by_kbid(kbid)
+            except ShardsNotFound:
+                logger.warning(
+                    "KB not found while looking for orphan shards", extra={"kbid": kbid}
+                )
+                continue
+            else:
+                for shard_object_pb in kb_shards:
+                    for shard_replica_pb in shard_object_pb.replicas:
+                        shard_replica_id = shard_replica_pb.shard.id
+                        node_id = shard_replica_pb.node
+                        stored_shards[shard_replica_id] = ShardLocation(
+                            kbid=kbid, node_id=node_id
+                        )
+    return stored_shards
+
+
+async def report_orphan_shards(
+    shards: dict[str, ShardLocation],
+    driver: Driver,
+):
+    async with driver.transaction(read_only=True) as txn:
+        for shard_id, location in shards.items():
+            kb_exists = await KnowledgeBox.exist_kb(txn, location.kbid)
+            if kb_exists:
+                msg = "Found orphan shard for existing KB"
+            else:
+                msg = "Found orphan shard for already removed KB"
+
+            logger.debug(
+                msg,
+                extra={
+                    "shard_id": shard_id,
+                    "kbid": location.kbid,
+                    "node_id": location.node_id,
+                },
+            )
+
+
+async def purge_orphan_shards(driver: Driver):
+    orphan_shards = await detect_orphan_shards(driver)
+
+    for shard_id, location in orphan_shards.items():
+        node = manager.get_index_node(location.node_id)
+        extra = {
+            "shard_id": shard_id,
+            "kbid": location.kbid,
+            "node_id": location.node_id,
+        }
+        if node is None:
+            logger.warning(
+                "Node not available while purging orphan shard, skipping", extra=extra
+            )
+            continue
+
+        logger.info("Deleting orphan shard from index node", extra=extra)
+        await node.delete_shard(shard_id)
+
+
+async def main():
+    """This script will detect orphan shards, i.e., indexed shards with no
+    reference in our source of truth.
+
+    Purging a knowledgebox can lead to orphan shards if index nodes are not
+    available or fail. It's possible that other procedures in our database left
+    orphan shards too.
+
+    ATENTION!
+    While migrations are running, new shards are indexed and are
+    not yet stored as shards belonging to a KB, so they can be misclassified as
+    orphan. It is highly recommended to don't remove orphan shards that hasn't
+    exist for a long time.
```

Comment:
Maybe you could extend the logic to detect them to check which shards are part of a reindex?